File: app/src/main/kotlin/com/spacex/ui/rockets/RocketsModule.java
package com.spacex.ui.rockets;

import dagger.Binds;
import dagger.Module;
import dagger.Provides;

@Module
public abstract class RocketsModule {

  @Provides
  static RocketsContract.Presenter providePresenter(RocketsContract.View view) {
    return new RocketsPresenter(view);
  }

  @Binds
  abstract RocketsContract.View bindView(RocketsActivity rocketsActivity);
}

File: app/src/main/kotlin/com/spacex/ui/rockets/RocketsModule.kt
package com.spacex.ui.rockets

import com.spacex.ui.rockets.RocketsActivity.Companion.ROCKET_IDS
import com.spacex.ui.rockets.RocketsActivity.Companion.SELECTED_ROCKET_ID
import dagger.Binds
import dagger.Module
import dagger.Provides

@Module
internal abstract class RocketsModule {

    @Binds
    abstract fun bindView(rocketsActivity: RocketsActivity): RocketsContract.View

    @Module
    companion object {

        @JvmStatic
        @Provides
        fun providePresenter(
                view: RocketsContract.View
        ): RocketsContract.Presenter {
            return RocketsPresenter(
                    view,
                    (view as RocketsActivity).intent.getStringExtra(SELECTED_ROCKET_ID),
                    view.intent.getStringArrayListExtra(ROCKET_IDS)
            )
        }
    }
}
